module/executiondatasync/optimistic_syncing/core.go

```diff
@@ -16,4 +28,208 @@ type Core interface {
+// It downloads execution data for the block and stores it in the execution data cache.
+// Expected errors:
+// - context.Canceled: if the provided context was canceled before completion
+//
+// No other errors are expected during normal operations
+func (c *CoreImpl) Download(ctx context.Context) error {
+	blockID := c.executionResult.BlockID
+	height := c.header.Height
+
+	c.log.Debug().
+		Hex("block_id", logging.ID(blockID)).
+		Uint64("height", height).
+		Msg("downloading execution data")
+
+	executionData, err := c.execDataRequester.RequestExecutionData(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to request execution data: %w", err)
+	}
+
+	c.executionData = execution_data.NewBlockExecutionDataEntity(c.executionResult.ExecutionDataID, executionData)
+
+	txResultErrMsgsData, err := c.txResultErrMsgsRequester.Request(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to request transaction result error messages data: %w", err)
+	}
+
+	c.txResultErrMsgsData = txResultErrMsgsData
+
+	return nil
+}
+
+// Index implements the Core.Index method.
+// It retrieves the downloaded execution data from the cache and indexes it into in-memory storage.
+// Expected errors:
+// - storage.ErrHeightNotIndexed if the given height does not match the storage's block height.
+//
+// No other errors are expected during normal operations
+func (c *CoreImpl) Index(_ context.Context) error {
+	c.log.Debug().Msg("indexing execution data")
+
+	if err := c.indexer.IndexBlockData(c.executionData); err != nil {
+		return err
+	}
+
+	if err := c.indexer.IndexTxResultErrorMessagesData(c.txResultErrMsgsData); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Persist implements the Core.Persist method.
+// It persists the indexed data to permanent storage atomically.
+//
+// No errors are expected during normal operations
+func (c *CoreImpl) Persist(_ context.Context) error {
+	c.log.Debug().Msg("persisting execution data")
+
+	// Add all data to the batch
+	if err := c.persister.Persist(); err != nil {
+		return fmt.Errorf("failed to persist data: %w", err)
+	}
+
+	c.log.Info().Msg("successfully persisted execution data")
+
+	return nil
+}
+
+// Abandon indicates that the protocol has abandoned this state. Hence processing will be aborted
+// and any data dropped.
+//
+// No errors are expected during normal operations
+func (c *CoreImpl) Abandon(_ context.Context) error {
+	c.log.Debug().Msg("Abandon execution data processing")
+
+	// Clear in-memory storage by setting references to nil for garbage collection
+	// Since we don't have Clear() methods, we remove the references to allow GC
+	c.inmemRegisters = nil
+	c.inmemEvents = nil
+	c.inmemCollections = nil
+	c.inmemTransactions = nil
+	c.inmemResults = nil
+	c.inmemTxResultErrMsgs = nil
+
+	// Clear other references
+	c.txResultErrMsgsRequester = nil
+	c.execDataRequester = nil
+	c.indexer = nil
+	c.persister = nil
+	c.executionData = nil
+	c.txResultErrMsgsData = nil
```

Comment:
consider putting all of these into a struct, then just setting that wrapper struct to nil. that will make it less likely we forget to add something in the future.